### ISSUE TITLE
The best camera setup for fine-tuning SmolVLA with so101

### ISSUE THREAD

**Issue by yuanlin2004** [huggingface/lerobot#1763]
I am fine-tuning SmolVLA for the pick-and-place task for my so101. What is the best camera setup? And what is the order to specify them for record.py?

I find the following pieces of information, which do not give a coherent answer. 

1. The SmolVLA [paper](https://arxiv.org/pdf/2506.01844) says 'prioritizing top, wrist, and side perspectives–and renamed them as OBS_IMAGE_1, OBS_IMAGE_2, and OBS_IMAGE_3'. 

2. This dataset from Huggingface for so101 [lerobot/svla\_so101\_pickplace · Datasets at Hugging Face](https://huggingface.co/datasets/lerobot/svla_so101_pickplace) uses 'up' and 'side'.

3. This dataset from Huggingface for so100 [lerobot/svla\_so101\_pickplace · Datasets at Hugging Face](https://huggingface.co/datasets/lerobot/svla_so101_pickplace) uses 'top' and 'wrist'.
 
4. The SmoVLA code ([prepare_images] (https://github.com/huggingface/lerobot/blob/55198de096f46a8e0447a8795129dd9ee84c088c/src/lerobot/policies/smolvla/modeling_smolvla.py#L479) and [embed_prefix] (https://github.com/huggingface/lerobot/blob/55198de096f46a8e0447a8795129dd9ee84c088c/src/lerobot/policies/smolvla/modeling_smolvla.py#L692) ) simply concatenates the image tokens from all the cameras (in the order specified in the command line argument of record.py) together with language tokens and state tokens as the input to SmolVLM. It does not matter how many cameras are used, as long as there is at least one. The names of the cameras do not matter either. There seems no clear way to tell from the embedding input which token is for which image or text. 
 
Is keeping the semantics of the embedding input in the fine-tuning data consistent with that in the pre-training data important for fine-tuning?

In my experiment, my fine-tuning [dataset](https://huggingface.co/spaces/lerobot/visualize_dataset?path=%2FTangyKiwi%2Frecord-test%2Fepisode_0) used 'top', 'front' and 'wrist' cameras and in that order. I found I got better result during inference when I keep the same order (i.e. `--robot.cameras="{ top: {..},  front: {..},  wrist: {..}}`, and worse result if different order (e.g. `--robot.cameras="{ top: {..},  wrist: {..},  front: {..}}`). But I do not know my camera order is consistent with the pre-training data. 

Could someone familiar with SmoVLA and code give an authoritative answer?

**Comment by yuanlin2004** [comment]
Just to answer my own question, the datasets used to pre-train the SmolVLA model can be found in the appendix A of the SmolVLA paper. I randomly inspected a few datasets and found they used very different camera views.

**Comment by MyLovelyAxe** [comment]
> Just to answer my own question, the datasets used to pre-train the SmolVLA model can be found in the appendix A of the SmolVLA paper. I randomly inspected a few datasets and found they used very different camera views.

Hi @yuanlin2004 , I just started to work in fine-tune SmolVLA recently. I am using it for a simulated so100 in Isaac Sim, no hardware, the input images and joint states are all from Isaac Sim. I found that the zero-shot SmolVLA always gave strange behavior, so fine-tunning might be a must. But I don't find any dataset for Isaac Sim scenarios, so I probably need to generate some dataset with Isaac Sim by myself. So I would appreciate to have some advices about the dataset itself, e.g.:

1) zero-shot

For now I haven't fine-tune SmolVLA, I just used the base model: 

`model = SmolVLAPolicy.from_pretrained("lerobot/smolvla_base")`

The robot arm always got stuck at a fixed position, is this the expected behavior?

2) camera poses

The dataset you mentioned in your question [lerobot/svla_so101_pickplace · Datasets at Hugging Face](https://huggingface.co/datasets/lerobot/svla_so101_pickplace) mentions `observation.images.up`, is this the same pose with what you called `top` later? Should the camera names MUST be called `up`, `side`, `front` instead of any other names like ´camera1´? And are there any evidence about these camera poses, i.e. what is the position and orientation of these `up`, `side`, `front`? Can cameras with any other poses be used to fine-tune?

**Comment by yuanlin2004** [comment]
@MyLovelyAxe I hope someone from the SmolVLA team can give answer the questions. From my reading of the code and the pre-training dataset, they seem to over all over the places - which may indicate the labeling, positions and ordering of the cameras do not matter.

**Comment by MyLovelyAxe** [comment]
> [@MyLovelyAxe](https://github.com/MyLovelyAxe) I hope someone from the SmolVLA team can give answer the questions. From my reading of the code and the pre-training dataset, they seem to over all over the places - which may indicate the labeling, positions and ordering of the cameras do not matter.

Hi thanks for replying! I have a question about the expected input features of SmolVLA. In a tutorial .py in the lerobot repo, i.e.
[lerobot/examples/tutorial/using_smolvla_example.py](https://github.com/huggingface/lerobot/blob/ce348a34607e2c9120fcd38b9ebfe1b37079c03c/examples/tutorial/smolvla/using_smolvla_example.py#L36), it mentions this:

`You can check the camera keys expected by a model in the info.json card on the model card on the Hub`

the only [info.json](https://huggingface.co/lerobot/smolvla_base/blob/main/config.json#L11) I can find has feature definition for images like this:


```
        "observation.images.camera1": {
            "type": "VISUAL",
            "shape": [
                3,
                256,
                256
            ]
        },
```

this is totally different from the [lerobot/svla_so101_pickplace](https://huggingface.co/datasets/lerobot/svla_so101_pickplace) or [lerobot/svla_so100_pickplace](https://huggingface.co/datasets/lerobot/svla_so100_pickplace), where:

```
        "observation.images.up": {
            "dtype": "video",
            "shape": [
                480,
                640,
                3
            ],
            "names": [
                "height",
                "width",
                "channels"
            ],
```

Which one did you use for fine-tuning tho? And does it improve the performance?

**Comment by yuanlin2004** [comment]
I used top/front/wrist. The performance was not good.

**Comment by MyLovelyAxe** [comment]
> I used top/front/wrist. The performance was not good.

Did you train for the so100 or so101? With the [training script](https://huggingface.co/docs/lerobot/il_robots#train-a-policy) offered by lerobot, i.e.
```

lerobot-train \
  --dataset.repo_id=${HF_USER}/so101_test \
  --policy.type=act \
  --output_dir=outputs/train/act_so101_test \
  --job_name=act_so101_test \
  --policy.device=cuda \
  --wandb.enable=true \
  --policy.repo_id=${HF_USER}/my_policy

```

**Comment by yuanlin2004** [comment]
I was using https://github.com/huggingface/notebooks/blob/main/lerobot/training-smolvla.ipynb

**Comment by mcres** [comment]
> The robot arm always got stuck at a fixed position, is this the expected behavior?

@MyLovelyAxe I also reproduced the same behavior you described, using the model without fine-tuning. Something similar was also reproduced by @alexcbb [here](https://huggingface.co/blog/Beegbrain/daily-robotics-june-2#a-zero-shot-tests-early-friction).

I think this comment will help: https://github.com/huggingface/lerobot/issues/2221#issuecomment-3413221473

**Comment by MyLovelyAxe** [comment]
> > The robot arm always got stuck at a fixed position, is this the expected behavior?
> 
> [@MyLovelyAxe](https://github.com/MyLovelyAxe) I also reproduced the same behavior you described, using the model without fine-tuning. Something similar was also reproduced by [@alexcbb](https://github.com/alexcbb) [here](https://huggingface.co/blog/Beegbrain/daily-robotics-june-2#a-zero-shot-tests-early-friction).
> 
> I think this comment will help: [#2221 (comment)](https://github.com/huggingface/lerobot/issues/2221#issuecomment-3413221473)

I think I found the reason of the issue that the robot arm always got stuck at the same position and kept wobbling. If SmolVLA base model is loaded from pre-trained model by:

```
from lerobot.policies.smolvla.modeling_smolvla import SmolVLAPolicy
model = SmolVLAPolicy.from_pretrained("lerobot/smolvla_base")
```

then according to the source code, the observation input is not normalized and action output is not denormalized. This is caused by a mis-match of label names in processor `safetensors` file and dataset structure. Check source code in [lerobot/src/lerobot/processor/normalize_processor.py](https://github.com/huggingface/lerobot/blob/9701b9c2735a989a6f07b5b3128db73408245c72/src/lerobot/processor/normalize_processor.py#L306):

```
        if norm_mode == NormalizationMode.IDENTITY or key not in self._tensor_stats:
            return tensor
```

The variable `key` here is string `"action"` defined in this file, which is also the dataset feature defined in `config.json`. But the `self._tensor_stats` is loaded from these `safetensors` files (since the policy was loaded by functioin `from_pretrained()`), check your downloaded pretrained model here:

```
~/.cache/huggingface/hub/models--lerobot--smolvla_base/snapshots/4d2f2b37fa245361ef1efe6d91ce96b8bd4af511/policy_postprocessor_step_0_unnormalizer_processor.safetensors
~/.cache/huggingface/hub/models--lerobot--smolvla_base/snapshots/4d2f2b37fa245361ef1efe6d91ce96b8bd4af511/policy_preprocessor_step_5_normalizer_processor.safetensors
```

These 2 `safetensors` are the ones got loaded into the `self._tensor_stats` above, if you load them with:

```
from safetensors.torch import load_file
from pathlib import Path
folder = Path("/home/hardli/.cache/huggingface/hub/models--lerobot--smolvla_base/snapshots/4d2f2b37fa245361ef1efe6d91ce96b8bd4af511")
postprocessor_safetensor = "policy_postprocessor_step_0_unnormalizer_processor.safetensors"
preprocessor_safetensor = "policy_preprocessor_step_5_normalizer_processor.safetensors"
post_processor = load_file(folder / postprocessor_safetensor)
pre_processor = load_file(folder / preprocessor_safetensor)
```

You will find the key names don't have a `action`:

```
{'so100-blue.buffer.action.mean': tensor([   1.4212,  125.7205,  125.4193,   63.7832, -106.5401,    3.3485]),
 'so100-blue.buffer.action.std': tensor([14.0820, 39.4314, 18.2381, 23.4898, 38.1112,  5.0483]),
 'so100-red.buffer.action.mean': tensor([   2.4387,  124.9379,  123.5578,   66.9926, -103.8768,    2.8912]),
 'so100-red.buffer.action.std': tensor([14.2992, 42.7722, 18.8763, 21.6519, 33.5341,  4.4780]),
 'so100.buffer.action.mean': tensor([  1.5959, 119.9442, 109.7698,  56.7055, -27.4227,  12.0028]),
 'so100.buffer.action.std': tensor([26.3918, 52.4114, 49.8538, 36.9983, 59.3600, 19.0400])}
```

and according to this if-condition mentioend above:

```
        if norm_mode == NormalizationMode.IDENTITY or key not in self._tensor_stats:
            return tensor
```

the `key not in self._tensor_stats` is always `True`, and the input `tensor` is just returned without any change. This means input is directly sent into policy without normalization, and output was sent out from policy without denormalization. Since the poliy itself was trained by normalized data, if without normalization, any input is not the same distribution that the policy expects. This causes the behavior that robot arm always wobbles at the same position, because the input and output are not correctly dealt with. However, as you see, this mismatch comes from:

- pre-trained safetensors for processors
- dataset feature (i.e. key name `"action"`) that SmolVLA was trained on

but since that if-condition just lets the input tensor returned without raising any error, even if it doesn't find correct key name in `self._tensor_stats`, it is hard to find whether the input and out are really normalized.

You can fix this temporarily by changing the key names in safetensors or dataset structure to let it match, afterwards, the robot should not get stuck anymore, even if zero-shot performance is not ideal, it looks more reasonable. But probably this still needs an official fix from Lerobot repo.

**Comment by devika-dudo** [comment]
hiii... one doubt...i read somehwere that if i am able to teleoperate properly looking only at the camera feed...then the camera setup is fine
is this correct??
